PX4FMU-V4PRO support

klopezal opened the pull request:
This pull request brings changes for px4fmu-v4pro : 

- correct RAM size for new processor (critical)
- correct GPIO logic for power supply sensing
- new chip select for temperature calibration EEPROM
- changes in the LIS3MDL driver

To be discussed : "artificial" rotations have been removed from the LIS3MDL driver. The aim here is to only use startup script with the "-R" option in order to correct the orientation between sensor's own orientation and autopilot's orientation.

davids5 commented on nuttx-configs/px4fmu-v4pro/nsh/defconfig:
```diff
@@ -627,7 +627,7 @@ CONFIG_BOOT_RUNFROMFLASH=y
 # Boot Memory Configuration
 #
 CONFIG_RAM_START=0x20000000
-CONFIG_RAM_SIZE=327680
```
FYI: These are not use in Nuttx on this Soc - I commited the fix upstream in [Nuttx here](https://bitbucket.org/nuttx/nuttx/commits/df9c5b33a01cac9da1786ede3903c575861cc769)

klopezal replied:
@davids5 
Actually the parameter is used in load monitor : https://github.com/PX4/Firmware/blob/master/src/modules/load_mon/load_mon.cpp#L243-L245

If the parameter is not right, it triggers an overload alert in commander : https://github.com/PX4/Firmware/blob/master/src/modules/commander/commander.cpp#L3137

Do you agree?

davids5 replied:
We need to see if there is a better way to do this, so there is only one source. I will open an issue.

davids5 commented on src/drivers/boards/px4fmu-v4pro/board_config.h:
```diff
@@ -95,10 +95,11 @@ __BEGIN_DECLS
 
 /* SPI Bus 1 Internal Sensors */
 
-#define GPIO_SPI_CS_MPU9250     (GPIO_OUTPUT|GPIO_PUSHPULL|GPIO_SPEED_2MHz|GPIO_OUTPUT_SET|GPIO_PORTC|GPIO_PIN2)
```
Can you restore the alignment so the diff is not so noisy?

LorenzMeier replied:
I think a new alignment was necessary due to tempcal EEPROM

davids5 commented on src/drivers/boards/px4fmu-v4pro/board_config.h:
```diff
@@ -330,7 +328,7 @@ __BEGIN_DECLS
 #define BOARD_ADC_SERVO_VALID   (1)
 #define BOARD_ADC_PERIPH_5V_OC  (px4_arch_gpioread(GPIO_VDD_5V_PERIPH_OC))
-#define BOARD_ADC_HIPOWER_5V_OC (0)
+#define BOARD_ADC_HIPOWER_5V_OC (px4_arch_gpioread(GPIO_VDD_5V_HIPOWER_OC))
```
Is this active low?

klopezal replied:
Yes, it has an external pull-up and low means there is an overvoltage/overcurrent condition.

davids5 replied:
We should check but I would expect that BOARD_ADC_HIPOWER_5V_OC  is positive true logic and 0 is not overcurrent. This is my assumption as when we do not have the signal a 0 was returned.

So wouldn't this be  #define BOARD_ADC_HIPOWER_5V_OC (!px4_arch_gpioread(GPIO_VDD_5V_HIPOWER_OC))

klopezal replied:
I think you are absolutely right, I will check.

klopezal replied:
I made the change, it makes sense. For the moment both 
`BOARD_ADC_HIPOWER_5V_OC ` and `BOARD_ADC_PERIPH_5V_OC ` are not used in [commander](https://github.com/PX4/Firmware/blob/master/src/modules/commander/commander.cpp#L1967-L2000), they appear only in [sdlog2](https://github.com/PX4/Firmware/blob/master/src/modules/sdlog2/sdlog2.c#L2099-L2115) for logging purposes. We should also add logic for the second brick, would that be better in a different PR?

LorenzMeier commented on src/drivers/lis3mdl/lis3mdl.cpp:
```diff
@@ -1118,63 +1138,41 @@ int LIS3MDL::calibrate(struct file *filp, unsigned enable)
 		warnx("FAILED: MAGIOCSRANGE 4 Ga");
 	}
-
+	
```
whitespace error

LorenzMeier commented on src/drivers/lis3mdl/lis3mdl.cpp:
```diff
@@ -1212,7 +1210,7 @@ int LIS3MDL::check_offset()
 
 	/* return 0 if calibrated, 1 else */
-	return !offset_valid;
+	return offset_valid;
```
This change is incorrect - this is returning an int, not a bool.

LorenzMeier commented on src/drivers/lis3mdl/lis3mdl.cpp:
```diff
@@ -1194,16 +1192,16 @@ int LIS3MDL::check_scale()
 
 	/* return 0 if calibrated, 1 else */
-	return !scale_valid;
+	return scale_valid;
```
This change is incorrect - confusion of int (0 == OK) and bool

LorenzMeier replied:
@klopezal Sorry, the fix you applied is still incorrect.

```
int ret = 1; // Error
int ret = 0; // OK

bool bool_ret = true; // OK
bool bool_ret = false; // Error

// conversion from bool to error code:
int ret = !bool_ret;
```

LorenzMeier commented:
Jenkins test this please.

LorenzMeier commented:
@klopezal Code style failures: https://travis-ci.org/PX4/Firmware/jobs/195933698

klopezal commented:
@LorenzMeier @davids5 
Corrected and squashed.

David, in accordance with my comment above I left Nuttx configuration as is, can you confirm?

Lorenz, FYI, there is the same error in the HMC5883 driver (confusion between int and bool), should I open a PR for this as well?

davids5 commented:
@klopezal  - that is fine for now - I did open an issue.

LorenzMeier commented:
@klopezal I don't think the HMC5883 has an issue. I rather think we had a miscommunication about the int / bool review comment I had.

klopezal commented:
@LorenzMeier sorry the confusion was on my side, I thought `FLT_EPSILON` was used to determine valid offset and scale thresholds while it is all the other way round. Offset thresholds are correctly set inside [calibration routine](https://github.com/PX4/Firmware/blob/master/src/modules/commander/mag_calibration.cpp#L71).

dagar commented on src/drivers/lis3mdl/lis3mdl.cpp:
```diff
@@ -1178,38 +1178,38 @@ int LIS3MDL::calibrate(struct file *filp, unsigned enable)
 
 int LIS3MDL::check_scale()
```
return bool?

klopezal replied:
Isn't that equivalent? Same in [HMC5883](https://github.com/PX4/Firmware/blob/master/src/drivers/hmc5883/hmc5883.cpp#L1253-L1269) and [IST8310](https://github.com/PX4/Firmware/blob/master/src/drivers/ist8310/ist8310.cpp#L1193-L1209) Should I change all of them?

dagar replied:
There's potentially a bit of ambiguity mixing int and bool return values, but it's not wrong.

dagar commented on src/drivers/lis3mdl/lis3mdl.cpp:
```diff
@@ -1178,38 +1178,38 @@ int LIS3MDL::calibrate(struct file *filp, unsigned enable)
+	return !scale_valid;
 }
 
 int LIS3MDL::check_offset()
```
return bool
